src/modules/cpu/kernel/spectrogram.hpp

```diff
@@ -0,0 +1,388 @@
+#include "rppdefs.h"
+#include "rpp_cpu_simd.hpp"
+#include "rpp_cpu_common.hpp"
+#include <chrono>
+#include <complex>
+
+inline Rpp32f reduce_add_ps1(__m256 src)
+{
+    __m256 src_add = _mm256_add_ps(src, _mm256_permute2f128_ps(src, src, 1));
+    src_add = _mm256_add_ps(src_add, _mm256_shuffle_ps(src_add, src_add, _MM_SHUFFLE(1, 0, 3, 2)));
+    src_add = _mm256_add_ps(src_add, _mm256_shuffle_ps(src_add, src_add, _MM_SHUFFLE(2, 3, 0, 1)));
+    Rpp32f *addResult = (Rpp32f *)&src_add;
+    return addResult[0];
+}
+
+inline void hann_window(Rpp32f *output, Rpp32s windowSize)
+{
+    Rpp32f a = (PI / windowSize);
+    for (Rpp32s t = 0; t < windowSize; t++)
+    {
+        Rpp32f phase = a * (t + 0.5);
+        output[t] = (0.5 * (1.0 - std::cos(phase)));
+    }
+}
+
+inline Rpp32s get_num_windows(Rpp32s length, Rpp32s windowLength, Rpp32s windowStep, bool centerWindows)
+{
+    if (!centerWindows)
+        length -= windowLength;
+
+    return ((length / windowStep) + 1);
+}
+
+inline Rpp32s get_idx_reflect(Rpp32s idx, Rpp32s lo, Rpp32s hi)
+{
+    if (hi - lo < 2)
+        return hi - 1;
+    for (;;)
+    {
+        if (idx < lo)
+            idx = 2 * lo - idx;
+        else if (idx >= hi)
+            idx = 2 * hi - 2 - idx;
+        else
+            break;
+    }
+    return idx;
+}
+
+RppStatus spectrogram_host_tensor(Rpp32f *srcPtr,
+                                  RpptDescPtr srcDescPtr,
+                                  Rpp32f *dstPtr,
+                                  RpptDescPtr dstDescPtr,
+                                  Rpp32s *srcLengthTensor,
+                                  bool centerWindows,
+                                  bool reflectPadding,
+                                  Rpp32f *windowFunction,
+                                  Rpp32s nfft,
+                                  Rpp32s power,
+                                  Rpp32s windowLength,
+                                  Rpp32s windowStep,
+                                  RpptSpectrogramLayout layout)
+{
+    Rpp32s windowCenterOffset = 0;
+    bool vertical = (layout == RpptSpectrogramLayout::FT);
+
+    if (centerWindows)
+        windowCenterOffset = windowLength / 2;
+
+    if (nfft == 0.0f)
+        nfft = windowLength;
+
+    Rpp32s numBins = nfft / 2 + 1;
+
+    std::vector<Rpp32f> windowFn;
+    windowFn.resize(windowLength);
+    // Generate hanning window
+    if (windowFunction == NULL) {
+        hann_window(windowFn.data(), windowLength);
+    } else {
+        memcpy(windowFn.data(), windowFunction, windowLength*sizeof(Rpp32f));
+    }
+
+    const Rpp32f mulFactor = (PI) / nfft;
+
+    Rpp32f cosf[numBins * nfft];
+    Rpp32f sinf[numBins * nfft];
+
+    for (Rpp32s k = 0; k < numBins; k++)
+    {
+        for (Rpp32s i = 0; i < nfft; i++)
+        {
+            cosf[k * nfft + i] = std::cos(k * i * mulFactor);
```

Comment:
Please try to vectorize this loop using [sincos_ps](https://github.com/sampath1117/rpp/blob/master/src/include/cpu/rpp_cpu_simd.hpp#L1705)

It will return the sin and cos value for the given input
You can refer below commit for the changes needed
https://github.com/sampath1117/rpp/commit/fae91471c81ff2bbda036056b863440215561bc4#diff-c088da125ac78f80dceb6ea17bce35c5bd4e526ac56e7cb474ae225af818ccf4

Reply:
Done